crates/worker/src/partition/state_machine/mod.rs

```diff
@@ -522,55 +575,16 @@ impl<Codec: RawEntryCodec> StateMachine<Codec> {
         Self::send_submit_notification_if_needed(
             ctx,
             service_invocation.invocation_id,
-            false,
+            // is_new_invocation is true if the RPC ingress request is a duplicate.
+            service_invocation.source
+                == *previous_invocation_status
+                    .source()
+                    .expect("source must be present when InvocationStatus is not Free"),
             service_invocation.submit_notification_sink,
         );
```

Comment:
I have to admit that I have a hard time following the flow of the `on_service_invocation` call and the methods it calls. Therefore, it's hard for me to assert the correctness. One problem that I see is that we send messages at different places (e.g. the submit notification) and based on some return value expect the caller to continue or not. Due to this property, one needs to follow every branch and check that all the required actions happen and when popping the stack on needs to remember based on the return value that certain things need to happen or not. 

One idea that could simplify things a bit is to do the I/O at one place (I guess in the `on_service_invocation` method).

Reply:
For now I'm not gonna take any action regarding this, I plan in future to reorganize this code a bit.